### ISSUE TITLE
Remake forkable reality.eth for L1 governance decisions

### ISSUE THREAD

**Issue by edmundedgar** [RealityETH/subjectivocracy#202]
We currently expect the L1 forkable reality.eth to only be used for contract upgrade decisions, and gated to the Security Council. These should not happen frequently and should generally not happen in parallel. However, it is possible that they might, for example if we intend to make an upgrade, realize that there's something else we should address and decide to cancel the previous upgrade and replace it with a new one.

Changes from stock reality.eth:

 * Contract knows its owner who is able to freeze it (the forkingmanager)
 * Contract knows its parent
 * Contract can be frozen by its owner (the forkingmanager). This stops all write operations including withdrawals of existing questions, except for:
   * migrating tokens after the fork
   * withdrawing tokens that are already in the user's balance (ie the claim has been done but they haven't called `withdraw` so they are sitting in the user's `balanceOf`)
 * Anyone can call to migrate the tokens after the fork. [Or it's called automatically as part of the forking process?]
 * Anyone can import any question from the parent. Imported questions are the same as the original question, except that:
    * they have a new arbitrator (the new forkingmanager or related contract) 
    *  if they were not finalized or pending arbitration as of the date of freeze, their finalization time will be advanced to their normal challenge period after the fork. [or after they were imported?]
 * We disable `createTemplate` and use a single hard-coded template created in the constructor. This avoids the need to migrate templates. We will give this template an ID that does not clash with the templates usually created in the constructor.
 * If practical, importing will fire the log events it would have fired if the question had been created fresh. This will avoid the need to do new UI work to fetch them from the parent contract.

The parent forkingmanager (or related contract) will do the following when there's a fork:
 * Freeze the forkable reality.eth contract
 * Deploy a new reality.eth contract for each fork
 * Migrate the tokens to their child instances.
 
The child forkingmanager (or a related contract) will do the following after forking
 * If it forked over an L1 question, migrate that question to the child
 * If it forked over an L1 question, send the arbitration result for that question to its newly-created reality.eth. This will be either "yes" or "no" depending whether it is the "yes" fork or the "no" fork

Non-features

We won't make a feature to unwind a question that is no longer relevant and return the bonds. (We may make this later for the L2 version.) Instead we'll ask the question in a way that would make good-but-redundant upgrade requests answerable as "true", but they won't have any effect except for returning bonds. Phrase them something like "Is this a legitimate upgrade?" as opposed to "Should we upgrade to this?".

**Comment by josojo** [comment]
> We currently expect the L1 forkable reality.eth to only be used for contract upgrade decisions

I think we might also use it for this one: https://github.com/RealityETH/subjectivocracy/issues/189

>Changes from stock reality.eth:

I think its a great fit to also use the create_children() structure from the forkableStructure Contract. Then parent contract and forkingManger are automatically managed.

>Anyone can call to migrate the tokens after the fork. [Or it's called automatically as part of the forking process?]

I would say let's keep this one different, since it would increase the complexity of the forking. But no strong opinion. This can be changed later easily.

>if they were not finalized or pending arbitration as of the date of freeze, their finalization time will be advanced to their normal challenge period after the fork. [or after they were imported?]

I think after importing is more secure, as this gives all player more time to react, if importing didn't success early for some reason. But not strong opinion.

>If practical, importing will fire the log events it would have fired if the question had been created fresh. This will avoid the need to do new UI work to fetch them from the parent contract.

I think we could put in the fork for a new parsing of the UI. Should be only a day of work, or do you think its more?

> Instead we'll ask the question in a way that would make good-but-redundant upgrade requests answerable as "true", but they won't have any effect except for returning bonds. Phrase them something like "Is this a legitimate upgrade?" as opposed to "Should we upgrade to this?".

Sounds good.

**Comment by edmundedgar** [comment]
Edited the description to remove the ability to withdraw balance on the parent after the fork, as it would require us to keep track of how much of the money we needed to keep back instead of splitting it and handing it over to the child. Instead we'll import both the balances and the questions, and if you have unclaimed funds you can withdraw them from the children.

**Comment by edmundedgar** [comment]
Just to check: If a governance proposition on L1 results in a fork, is it OK to freeze consideration of any other governance proposition until the fork happens? @josojo 

If we don't do this it adds quite a bit of complexity because at a minimum we need our `L1ForkArbitrator` to be able to queue propositions and pass on its questions and propositions to a child, and also split its funds.
